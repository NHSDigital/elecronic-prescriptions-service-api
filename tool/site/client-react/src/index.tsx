import {Bundle} from "fhir/r4"
import * as React from "react"
import {
  createPrescriptionSummaryProps,
  PrescriptionSummary
} from "./components/prescription-summary/prescriptionSummary"
import {PageContainer} from "./components/pageContainer"
import ReactDOM = require("react-dom")

(async function() {
  const baseUrl = "/"
  const urlParams = new URLSearchParams(window.location.search)
  const prescriptionId = urlParams.get("prescription_id")
  const getPrescriptionResponse: Response = await fetch(`${baseUrl}prescription/${prescriptionId}`)
  const bundle: Bundle = await getPrescriptionResponse.json()
  const prescriptionSummaryProps = createPrescriptionSummaryProps(bundle)
  const content = (
    <PageContainer>
      <PrescriptionSummary
        patient={prescriptionSummaryProps.patient}
        practitionerRole={prescriptionSummaryProps.practitionerRole}
      />
    </PageContainer>
  )
  ReactDOM.render(content, document.getElementById("root"))
}())
