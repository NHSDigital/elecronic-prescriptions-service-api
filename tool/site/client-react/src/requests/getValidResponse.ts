import {AxiosResponse} from "axios"
import {OperationOutcome} from "fhir/r4"
import {UnhandledAxiosResponseError} from "./unhandledAxiosResponseError"

export function getResponseDataIfValid<T>(
  response: AxiosResponse<T | OperationOutcome>,
  typeGuard: (unknown: unknown) => unknown is T
): T {
  const responseData = response.data
  if (responseData && typeGuard(responseData)) {
    return responseData
  }

  throw new UnhandledAxiosResponseError("Unexpected response from server.", response)
}
