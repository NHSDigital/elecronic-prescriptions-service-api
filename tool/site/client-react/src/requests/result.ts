import {FhirResource} from "fhir/r4"

export function isResult(data: unknown): data is Result {
  const result = data as Result
  return typeof result.success === "boolean"
    && "request" in result
    && "request_xml" in result
    && "response" in result
    && "response_xml" in result
}

export interface Result {
  success: boolean
  request: FhirResource
  request_xml: string
  response: FhirResource
  response_xml: string
}
