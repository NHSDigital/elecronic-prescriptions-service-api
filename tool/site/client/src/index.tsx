import * as React from "react"
import {Page} from "./components/page"
import SendPreSignPage from "./pages/sendPreSignPage"
import * as ReactDOM from "react-dom"
import {BrowserRouter, Route, Switch} from "react-router-dom"
import ClaimPage from "./pages/claimPage"
import DispensePage from "./pages/dispensePage"
import {CookiesProvider} from "react-cookie"
import SendPostSignPage from "./pages/sendPostSignPage"
import HomePage from "./pages/homePage"
import PrescriptionSearchPage from "./pages/prescriptionSearchPage"
import ReleasePage from "./pages/releasePage"
import CancelPage from "./pages/cancelPage"
import MyPrescriptionsPage from "./pages/myPrescriptionsPage"
import LoginPage from "./pages/loginPage"
import ValidatePage from "./pages/validatePage"
import ReturnPage from "./pages/returnPage"
import WithdrawPage from "./pages/withdrawPage"
import LoadPage from "./pages/loadPage"
import {Environment} from "./services/environment"
import LogoutPage from "./pages/logoutPage"
import ConfigPage from "./pages/configPage"
import VerifyPage from "./pages/verifyPage"
import ComparePage from "./pages/comparePage"
import {PageFullWidth} from "./components/pageFullWidth"
import {PageContainer} from "./components/pageContainer"

const customWindow = window as Record<string, any>

export interface AppContextValue {
  baseUrl: string
  environment: Environment
}

export const AppContext = React.createContext<AppContextValue>({baseUrl: "/", environment: "int"})

async function startApplication(baseUrl: string, environment: Environment): Promise<void> {
  const urlParams = new URLSearchParams(window.location.search)
  const content = (
    <AppContext.Provider value={{baseUrl, environment}}>
      <CookiesProvider>
        <Page>
          <BrowserRouter>
            <Switch>
              <Route path={`${baseUrl}/`}>
                <PageContainer>
                  <HomePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}config`}>
                <PageContainer>
                  <ConfigPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}my-prescriptions`}>
                <PageContainer>
                  <MyPrescriptionsPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}login`}>
                <PageContainer>
                  <LoginPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}logout`}>
                <PageContainer>
                  <LogoutPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/load`}>
                <PageContainer>
                  <LoadPage/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/edit`}>
                <PageContainer>
                  <SendPreSignPage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/send`}>
                <PageContainer>
                  <SendPostSignPage token={urlParams.get("token")} state={urlParams.get("state")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}prescribe/cancel`}>
                <PageContainer>
                  <CancelPage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/release`}>
                <PageContainer>
                  <ReleasePage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/verify`}>
                <PageContainer>
                  <VerifyPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/return`}>
                <PageContainer>
                  <ReturnPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/withdraw`}>
                <PageContainer>
                  <WithdrawPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/dispense`}>
                <PageContainer>
                  <DispensePage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}dispense/claim`}>
                <PageContainer>
                  <ClaimPage prescriptionId={urlParams.get("prescription_id")}/>
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}search`}>
                <PageContainer>
                  <PrescriptionSearchPage prescriptionId={urlParams.get("prescription_id")} />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}validate`}>
                <PageContainer>
                  <ValidatePage />
                </PageContainer>
              </Route>
              <Route path={`${baseUrl}compare-prescriptions`}>
                <PageFullWidth>
                  <ComparePage />
                </PageFullWidth>
              </Route>
            </Switch>
          </BrowserRouter>
        </Page>
      </CookiesProvider>
    </AppContext.Provider>
  )
  ReactDOM.render(content, document.getElementById("root"))
}

customWindow.startApplication = startApplication
