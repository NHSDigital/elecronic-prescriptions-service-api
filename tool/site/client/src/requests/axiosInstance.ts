import axios from "axios"

export const axiosInstance = axios.create({
  validateStatus: () => true
})

// todo: refresh token if required
axiosInstance.interceptors.request.use(function (config) {
  return config
}, function (error) {
  return Promise.reject(error)
})
