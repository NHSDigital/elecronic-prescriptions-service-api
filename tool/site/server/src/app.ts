import Hapi from "@hapi/hapi"
import routes from "./routes"
import HapiPino from "hapi-pino"
import Vision from "@hapi/vision"
import * as inert from "@hapi/inert"
import Yar from "@hapi/yar"
import Cookie from "@hapi/cookie"
import {isDev, isLocal} from "./services/environment"
import axios from "axios"
import {setSessionValue} from "./services/session"

const init = async () => {
  axios.defaults.validateStatus = () => true

  const server = Hapi.server({
    port: 9000,
    host: "0.0.0.0",
    routes: {
      cors: true // Won't run as Apigee hosted target without this
    }
  })

  const baseUrl = process.env.BASE_PATH
    ? `/${process.env.BASE_PATH}/`
    : "/"

  await server.register(Cookie)
  server.auth.strategy("session", "cookie", {
    cookie: {
      name: "auth",
      password: process.env.SESSION_TOKEN_ENCRYPTION_KEY,
      isSecure: true
    },
    redirectTo: (request: Hapi.Request) => {
      if (isDev()) {
        setSessionValue(
          "use_signing_mock",
          request.query["use_signing_mock"],
          request
        )
      }
      return `${baseUrl}login`
    }
  })
  server.auth.default("session")

  server.route(routes)

  await server.register({
    plugin: Yar,
    options: {
      storeBlank: false,
      // Use "0" maxCookieSize to force all session data to be written to cache
      maxCookieSize: 0,
      cache: {
        expiresIn: 24 * 60 * 60 * 1000
      },
      cookieOptions: {
        password: process.env.SESSION_TOKEN_ENCRYPTION_KEY ?? "",
        isSecure: true,
        isSameSite: "None"
      }
    }
  })

  await server.register({
    plugin: HapiPino,
    options: {
      // Pretty print in local environment only to avoid spamming logs
      prettyPrint: isLocal(),
      // Redact Authorization headers, see https://getpino.io/#/docs/redaction
      redact: ["req.headers.authorization"]
    }
  })

  await server.register(inert)

  server.route({
    method: "GET",
    path: `/static/{param*}`,
    options: {
      auth: false
    },
    handler: {
      directory: {
        path: "static"
      }
    }
  })

  await server.register(Vision)

  server.views({
    engines: {
      html: require("handlebars")
    },
    relativeTo: __dirname,
    path: "templates"
  })

  server.route(addHomeViewRoute())
  server.route(addViewRoute("login", true))
  server.route(addViewRoute("my-prescriptions"))
  server.route(addViewRoute("validate"))
  server.route(addViewRoute("search"))
  server.route(addViewRoute("prescribe/load"))
  server.route(addViewRoute("prescribe/edit"))
  server.route(addViewRoute("prescribe/send"))
  server.route(addViewRoute("prescribe/cancel"))
  server.route(addViewRoute("dispense/release"))
  server.route(addViewRoute("dispense/return"))
  server.route(addViewRoute("dispense/dispense"))
  server.route(addViewRoute("dispense/withdraw"))
  server.route(addViewRoute("dispense/claim"))

  function addHomeViewRoute() : Hapi.ServerRoute {

    const baseUrl = process.env.BASE_PATH
      ? `/${process.env.BASE_PATH}/`
      : "/"

    return {
      method: "GET",
      path: `/`,
      handler: {
        view: {
          template: "index",
          context: {
            baseUrl,
            environment: process.env.ENVIRONMENT
          }
        }
      }
    }
  }

  function addViewRoute(path: string, skipAuth?: boolean): Hapi.ServerRoute {
    const baseUrl = process.env.BASE_PATH
      ? `/${process.env.BASE_PATH}/`
      : "/"

    const viewRoute = {
      method: "GET",
      path: `/${path}`,
      handler: {
        view: {
          template: "index",
          context: {
            baseUrl,
            environment: process.env.ENVIRONMENT
          }
        }
      }
    }

    if (skipAuth) {
      return {
        ...viewRoute,
        options: {
          auth: false
        }
      }
    }

    return viewRoute
  }

  await server.start()
  server.log("info", `Server running on ${server.info.uri}`)
}

process.on("unhandledRejection", err => {
  console.log(err)
  process.exit(1)
})

init()
