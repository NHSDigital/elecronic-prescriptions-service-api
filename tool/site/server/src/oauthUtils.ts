import ClientOAuth2 from "client-oauth2"
import {URL} from "url"
import {CONFIG} from "./config"
import {getRegisteredCallbackUrl} from "./routes/helpers"

export interface OAuthClient {
  getUri: (options?: ClientOAuth2.Options) => string
  getToken: (uri: URL) => Promise<Token>
}

export interface Token {
  data: ClientOAuth2.Data
  tokenType: string
  accessToken: string
  refreshToken: string
}

const oauthClient =
  new ClientOAuth2({
    clientId: CONFIG.clientId,
    clientSecret: CONFIG.clientSecret,
    redirectUri: getRegisteredCallbackUrl("callback"),
    accessTokenUri: `${CONFIG.privateApigeeUrl}/oauth2/token`,
    authorizationUri: `${CONFIG.publicApigeeUrl}/oauth2/authorize`,
    body: {
      client_id: CONFIG.clientId,
      client_secret: CONFIG.clientSecret
    }
  })

export default function getOAuthClient(): OAuthClient {
  return oauthClient.code
}

export async function refreshToken(data: ClientOAuth2.Data): Promise<ClientOAuth2.Token> {
  const oauthClientToken = oauthClient.createToken(data)
  const refreshedToken = await oauthClientToken.refresh()
  oauthClientToken.expiresIn(parseInt(oauthClientToken.data.refreshTokenExpiresIn))
  return refreshedToken
}
