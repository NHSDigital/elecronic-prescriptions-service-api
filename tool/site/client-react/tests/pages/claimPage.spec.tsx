import {waitFor} from "@testing-library/react"
import {screen} from "@testing-library/dom"
import pretty from "pretty"
import * as React from "react"
import moxios from "moxios"
import ClaimPage from "../../src/pages/claimPage"
import userEvent from "@testing-library/user-event"
import {readMessage} from "../messages/messages"
import {AppContextValue} from "../../src"
import {renderWithContext} from "../renderWithContext"
import {axiosInstance} from "../../src/requests/axiosInstance"

const baseUrl = "baseUrl/"
const prescriptionId = "7A9089-A83008-56A03J"
const context: AppContextValue = {baseUrl}

const prescriptionOrderUrl = `${baseUrl}prescription/${prescriptionId}`
const dispenseNotificationUrl = `${baseUrl}dispenseNotifications/${prescriptionId}`
const claimUrl = `${baseUrl}dispense/claim`

const prescriptionOrder = readMessage("prescriptionOrder.json")
const dispenseNotification = readMessage("dispenseNotification.json")

beforeEach(() => moxios.install(axiosInstance))

afterEach(() => moxios.uninstall(axiosInstance))

test("Displays loading text while prescription data is being requested", async () => {
  const {container} = renderWithContext(<ClaimPage prescriptionId={prescriptionId}/>, context)
  await waitFor(() => screen.getByText("Retrieving prescription details."))

  expect(screen.getByText("Loading...")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays claim form if prescription details are retrieved successfully", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 200,
    response: prescriptionOrder
  })
  moxios.stubRequest(dispenseNotificationUrl, {
    status: 200,
    response: [dispenseNotification]
  })

  const container = await renderPage()

  expect(screen.getByText("Claim")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays an error if prescription-order not found", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 200,
    response: null
  })

  const {container} = renderWithContext(<ClaimPage prescriptionId={prescriptionId}/>, context)
  await waitFor(() => screen.getByText("Error"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays an error if dispense-notification not found", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 200,
    response: prescriptionOrder
  })
  moxios.stubRequest(dispenseNotificationUrl, {
    status: 200,
    response: []
  })

  const {container} = renderWithContext(<ClaimPage prescriptionId={prescriptionId}/>, context)
  await waitFor(() => screen.getByText("Error"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays an error on invalid response", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 500,
    response: {}
  })

  const {container} = renderWithContext(<ClaimPage prescriptionId={prescriptionId}/>, context)
  await waitFor(() => screen.getByText("Error"))

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays loading text while claim is being submitted", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 200,
    response: prescriptionOrder
  })
  moxios.stubRequest(dispenseNotificationUrl, {
    status: 200,
    response: [dispenseNotification]
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Claim"))
  await waitFor(() => screen.getByText("Loading..."))

  expect(screen.getByText("Sending claim.")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test("Displays claim result", async () => {
  moxios.stubRequest(prescriptionOrderUrl, {
    status: 200,
    response: prescriptionOrder
  })
  moxios.stubRequest(dispenseNotificationUrl, {
    status: 200,
    response: [dispenseNotification]
  })
  moxios.stubRequest(claimUrl, {
    status: 200,
    response: {
      success: true,
      request: "JSON Request",
      request_xml: "XML Request",
      response: "JSON Response",
      response_xml: "XML Response"
    }
  })

  const container = await renderPage()
  userEvent.click(screen.getByText("Claim"))
  await waitFor(() => screen.getByText(/Claim Result/))

  expect(screen.getByText(JSON.stringify("JSON Request"))).toBeTruthy()
  expect(screen.getByText("XML Request")).toBeTruthy()
  expect(screen.getByText(JSON.stringify("JSON Response"))).toBeTruthy()
  expect(screen.getByText("XML Response")).toBeTruthy()
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

async function renderPage() {
  const {container} = renderWithContext(<ClaimPage prescriptionId={prescriptionId}/>, context)
  await waitFor(() => screen.getByText("Claim for Dispensed Medication"))
  return container
}
