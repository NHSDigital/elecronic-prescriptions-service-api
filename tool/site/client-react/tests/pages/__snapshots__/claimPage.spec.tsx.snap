// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays an error if dispense-notification not found 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Dispense notification not found. Has this prescription been dispensed?</span>
<div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"button\\">Back</button></div>"
`;

exports[`Displays an error if prescription-order not found 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Unexpected response from server.</span>
<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Status</dt>
    <dd class=\\"nhsuk-summary-list__value\\">200 </dd>
  </div>
</dl>
<details class=\\"nhsuk-details nhsuk-expander\\">
  <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response Headers</span></summary>
  <div class=\\"nhsuk-details__text\\">
    <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/plain;charset=utf-8,undefined\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\"></pre>
  </div>
</details>
<details class=\\"nhsuk-details nhsuk-expander\\">
  <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response Body</span></summary>
  <div class=\\"nhsuk-details__text\\">
    <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/plain;charset=utf-8,null\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">null</pre>
  </div>
</details>
<div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"button\\">Back</button></div>"
`;

exports[`Displays an error on invalid response 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Error</label></h1><span class=\\"nhsuk-error-message\\" role=\\"alert\\"><span class=\\"nhsuk-u-visually-hidden\\">Error: </span>Unexpected response from server.</span>
<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Status</dt>
    <dd class=\\"nhsuk-summary-list__value\\">500 </dd>
  </div>
</dl>
<details class=\\"nhsuk-details nhsuk-expander\\">
  <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response Headers</span></summary>
  <div class=\\"nhsuk-details__text\\">
    <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/plain;charset=utf-8,undefined\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\"></pre>
  </div>
</details>
<details class=\\"nhsuk-details nhsuk-expander\\">
  <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response Body</span></summary>
  <div class=\\"nhsuk-details__text\\">
    <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/plain;charset=utf-8,%7B%7D\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">{}</pre>
  </div>
</details>
<div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"button\\">Back</button></div>"
`;

exports[`Displays claim form if prescription details are retrieved successfully 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Claim for Dispensed Medication</label></h1>
<form>
  <div class=\\"nhsuk-form-group\\">
    <fieldset class=\\"nhsuk-fieldset\\">
      <legend class=\\"nhsuk-fieldset__legend nhsuk-fieldset__legend--m\\">Methotrexate 10mg/0.2ml solution for injection pre-filled syringes</legend>
      <dl class=\\"nhsuk-summary-list nhsuk-summary-list--no-border\\">
        <div class=\\"nhsuk-summary-list__row\\">
          <dt class=\\"nhsuk-summary-list__key\\">Status</dt>
          <dd class=\\"nhsuk-summary-list__value\\">Item fully dispensed</dd>
        </div>
        <div class=\\"nhsuk-summary-list__row\\">
          <dt class=\\"nhsuk-summary-list__key\\">Quantity Dispensed</dt>
          <dd class=\\"nhsuk-summary-list__value\\">63 tablet</dd>
        </div>
      </dl>
      <div class=\\"nhsuk-form-group\\">
        <div class=\\"nhsuk-checkboxes\\" id=\\"products.0.patientPaid.boxes\\">
          <div class=\\"nhsuk-checkboxes__item\\"><input class=\\"nhsuk-checkboxes__input\\" name=\\"products.0.patientPaid\\" id=\\"products.0.patientPaid.box\\" type=\\"checkbox\\" value=\\"false\\"><label class=\\"nhsuk-label nhsuk-checkboxes__label\\" id=\\"products.0.patientPaid.box--label\\" for=\\"products.0.patientPaid.box\\">Patient Paid</label></div>
        </div>
      </div>
      <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button\\" aria-disabled=\\"false\\" type=\\"button\\">Add Endorsement</button></div>
    </fieldset>
  </div>
  <div class=\\"nhsuk-form-group\\">
    <fieldset class=\\"nhsuk-fieldset\\">
      <legend class=\\"nhsuk-fieldset__legend nhsuk-fieldset__legend--m\\">Prescription Charge Exemption</legend>
      <div class=\\"nhsuk-form-group\\"><label class=\\"nhsuk-label\\" id=\\"exemption.code--label\\" for=\\"exemption.code\\">Exemption Status</label><select class=\\"nhsuk-select\\" aria-labelledby=\\"exemption.code--label\\" name=\\"exemption.code\\" id=\\"exemption.code\\">
          <option value=\\"0001\\">Patient has paid appropriate charges</option>
          <option value=\\"0002\\">is under 16 years of age</option>
          <option value=\\"0003\\">is 16, 17 or 18 and in full-time education</option>
          <option value=\\"0004\\">is 60 years of age or over</option>
          <option value=\\"0005\\">has a valid maternity exemption certificate</option>
          <option value=\\"0006\\">has a valid medical exemption certificate</option>
          <option value=\\"0007\\">has a valid prescription pre-payment certificate</option>
          <option value=\\"0008\\">has a War Pension exemption certificate</option>
          <option value=\\"0009\\">is named on a current HC2 charges certificate</option>
          <option value=\\"0010\\">was prescribed free-of-charge contraceptives</option>
          <option value=\\"0011\\">gets income support (IS)</option>
          <option value=\\"0012\\">gets income based Job Seeker's Allowance (JSA (IB))</option>
          <option value=\\"0013\\">is entitled to, or named on a VALID NHS tax credit exemption certificate</option>
          <option value=\\"0014\\">has a partner who gets Pension Credit Guarantee Credit (PGCC)</option>
          <option value=\\"0015\\">Patient does not need to pay the prescription charge</option>
        </select></div>
      <div class=\\"nhsuk-form-group\\">
        <div class=\\"nhsuk-checkboxes\\" id=\\"exemption.evidenceSeen.boxes\\">
          <div class=\\"nhsuk-checkboxes__item\\"><input class=\\"nhsuk-checkboxes__input\\" name=\\"exemption.evidenceSeen\\" id=\\"exemption.evidenceSeen.box\\" type=\\"checkbox\\" value=\\"false\\"><label class=\\"nhsuk-label nhsuk-checkboxes__label\\" id=\\"exemption.evidenceSeen.box--label\\" for=\\"exemption.evidenceSeen.box\\">Evidence Seen</label></div>
        </div>
      </div>
    </fieldset>
  </div>
  <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button\\" aria-disabled=\\"false\\" type=\\"submit\\">Claim</button><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"button\\">Back</button></div>
</form>"
`;

exports[`Displays claim result 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Claim Result <svg class=\\"nhsuk-icon nhsuk-icon__tick\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" height=\\"32\\" width=\\"32\\">
      <path stroke-width=\\"4\\" stroke-linecap=\\"round\\" stroke=\\"#007f3b\\" fill=\\"none\\" d=\\"M18.4 7.8l-8.5 8.4L5.6 12\\"></path>
    </svg></label></h1>
<div class=\\"nhsuk-action-link\\"><a class=\\"nhsuk-action-link__link\\" href=\\"baseUrl/search?prescription_id=7A9089-A83008-56A03J\\"><svg class=\\"nhsuk-icon nhsuk-icon__arrow-right-circle\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" height=\\"32\\" width=\\"32\\">
      <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
      <path d=\\"M12 2a10 10 0 0 0-9.95 9h11.64L9.74 7.05a1 1 0 0 1 1.41-1.41l5.66 5.65a1 1 0 0 1 0 1.42l-5.66 5.65a1 1 0 0 1-1.41 0 1 1 0 0 1 0-1.41L13.69 13H2.05A10 10 0 1 0 12 2z\\"></path>
    </svg><span class=\\"nhsuk-action-link__text\\">View prescription</span></a></div>
<div class=\\"nhsuk-expander-group\\">
  <details class=\\"nhsuk-details nhsuk-expander\\">
    <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Request (FHIR)</span></summary>
    <div class=\\"nhsuk-details__text\\">
      <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:application/json;charset=utf-8,%22JSON%20Request%22\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">\\"JSON Request\\"</pre>
    </div>
  </details>
  <details class=\\"nhsuk-details nhsuk-expander\\">
    <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Request (HL7 V3)</span></summary>
    <div class=\\"nhsuk-details__text\\">
      <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/xml;charset=utf-8,XML%20Request\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">XML Request</pre>
    </div>
  </details>
  <details class=\\"nhsuk-details nhsuk-expander\\">
    <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response (FHIR)</span></summary>
    <div class=\\"nhsuk-details__text\\">
      <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:application/json;charset=utf-8,%22JSON%20Response%22\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">\\"JSON Response\\"</pre>
    </div>
  </details>
  <details class=\\"nhsuk-details nhsuk-expander\\">
    <summary class=\\"nhsuk-details__summary\\"><span class=\\"nhsuk-details__summary-text\\">Response (HL7 V3)</span></summary>
    <div class=\\"nhsuk-details__text\\">
      <div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" aria-disabled=\\"false\\" type=\\"submit\\">Copy</button><a class=\\"nhsuk-button sc-dkPtRN hhyeAF\\" role=\\"button\\" aria-disabled=\\"false\\" draggable=\\"false\\" href=\\"data:text/xml;charset=utf-8,XML%20Response\\" download=\\"message\\">Download</a></div><pre class=\\"sc-gsDKAQ jsqwND\\">XML Response</pre>
    </div>
  </details>
</div>
<div class=\\"sc-bdvvtL jjZBkx\\"><button class=\\"nhsuk-button nhsuk-button--secondary\\" aria-disabled=\\"false\\" type=\\"button\\">Back</button></div>"
`;

exports[`Displays loading text while claim is being submitted 1`] = `"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Loading...</label></h1><label class=\\"nhsuk-label\\">Sending claim.</label>"`;

exports[`Displays loading text while prescription data is being requested 1`] = `"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Loading...</label></h1><label class=\\"nhsuk-label\\">Retrieving prescription details.</label>"`;
